### DISCUSSION TITLE
Notification System

### DISCUSSION THREAD

**Discussion by alexbrowng** [directus/directus#7394] [Feature Requests]
_This is a part of a series of proposals related to create a migration system, in which I would like to know your opinion and suggestions._

_My intention is, if it is interesting for the community and the internal team gives the go-ahead, I would start developing it myself._

_Sorry if I have duplicated another idea._

_Thanks in advance._

### Needs

A system that allows to register notifications to be read by users.

### Use cases

- An intervention has been scheduled.
- New application features.
- User reminders.
- Logging of account activity.
- Changes in a customer's subscription.

### Background

When doing a backup, maintenance tasks or migrating the data model, roles and permissions it is important to keep the users informed of the system status and its limitations or possible errors/inconsistencies, therefore, I have created a notification table where I indicate the type, subject, message and user filter (by role or by some field of the user data model) and with another table I relate the users and the notifications and keep a record of who has read it and when, in the intermediate table I only add the record when the user logs in or when he actively request his notifications (this is a custom endpoint), this way, I avoid, in case of having many users, to create many records when creating a notification, besides, some notifications are ephemeral, therefore, if the user accesses the platform and the notification is finished he does not receive that message.

### Proposal

Create a notification system in directus, on the one hand, you must create the data model, the service to manage them internally and the controller for the API, also, implement in the App a notification center and a new page where you can create them, as well as review the history of notifications, which users have received and which have read them.

### Minimal version

The initial version includes the option to register from the App a notification indicating the subject, the message, the rules for the users and the expiration date (optional). Allow in the text fields to add dynamic variables, for example: Hello, {{first_name}}....

**Data model**.

directus_notifications (id, created_at, updated_at, status (draft, published, archived), rules, expires_at (optional), subject, body)

directus_users_directus_notifications (id, created_at, user, notification, read_at)

directus_users (last_notifications_check)

**User service**

GET /users/me/notifications: Read user notifications, update last_notifications_check field to not check all system notifications on each request, only those created or updated after last check, if new notifications exist and user matches rules, create record in relational table.

PATCH /users/me/notifications/:notificationId: Allows to mark a notification as read.

**Notification service**

GET /notifications

GET /notifications/:notificationId

POST /notifications

PATCH /notifications/:notificationId

DELETE /notifications/:notificationId

### Futuribles

- Trigger system, send email on notification creation
- Types of notifications

**Comment by rijkvanzanten** [comment]
In favor 🙋🏻‍♂️

Like you said, this would also be really useful for other tasks that take time that don't happen in a single request, like generating an export of a collection with a bunch of items.

That being said, we should figure out how this ties into comments and activity. Instead of storing item comments under `directus_activity`, I think it would make more sense to tie that into this same "messaging" system, and maybe call it `directus_messages` instead (w/ `directus_messages_read` as junction table for users<->messages) 🤔

**Reply by benhaynes** [reply]
Thank you for the awesome info here, @alejandro-tss! Also adding this Discussion here, as it seems to have overlap: https://github.com/directus/directus/discussions/4002

**Reply by alexbrowng** [reply]
> Thank you for the awesome info here, @alejandro-tss! Also adding this Discussion here, as it seems to have overlap: https://github.com/directus/directus/discussions/4002

I had already read that discussion and I think they are complementary not so much as overlapping since, the alert system I understand is intended to make a notification in case of change of data in the database, in this case, my idea is more oriented to have a user messaging system that can be used for any purpose.

But I think that the two ideas together open a great world of possibilities.

**Reply by rijkvanzanten** [reply]
> But I think that the two ideas together open a great world of possibilities.

Yeah! Your proposal would be the place where these generated alerts are stored 🙂

**Reply by benhaynes** [reply]
Agree. Just trying to zoom out to see if there are good opportunities to re-use similar tables, with flags/scoping for different purposes. Clearly, we'll have to ensure that data belongs together before merging the tables though. 👍 

EDIT: I'm very OCD about not ending up with a million system tables... but maybe I'll have to relax that a bit! 😉

**Reply by rijkvanzanten** [reply]
> EDIT: I'm very OCD about not ending up with a million system tables... but maybe I'll have to relax that a bit! 😉

One more point for the "put the system tables in a different database/schema than the user tables" feature request 👀

**Reply by alexbrowng** [reply]
> EDIT: I'm very OCD about not ending up with a million system tables... but maybe I'll have to relax that a bit! wink

I agree with you, it is better not to have many system tables, but, it is better to separate the data by features, because, if not, you limit the evolution of the functionalities to the dependencies with the rest of the features.

**Comment by jamescammarano** [comment]
So the banner... goes on the directus app that users use? Or does it get pulled from the API and used on whatever front end you are making? (If it's available on the app it should be available through the API IMO)

Emailing wouldn't allow for an expiration date but would potentially be an easy hook into the recently implemented comment notifications.
